classes/phing/tasks/ext/phpunit/PHPUnitTestRunner9.php

```diff
@@ -0,0 +1,496 @@
+     * @param $message
+     * @param $file
+     * @param $line
+     * @return bool
+     */
+    public function handleError($level, $message, $file, $line): bool
+    {
+        $invoke = new PHPUnit\Util\ErrorHandler(true, true, true, true);
+        return $invoke($level, $message, $file, $line);
+    }
+
+    /**
+     * Run a test
+     *
+     * @param  PHPUnit\Framework\TestSuite $suite
+     * @throws \BuildException
+     * @throws ReflectionException
+     */
+    public function run(PHPUnit\Framework\TestSuite $suite)
+    {
+        $res = new PHPUnit\Framework\TestResult();
+
+        if ($this->codecoverage) {
+            $res->setCodeCoverage($this->codecoverage);
+        }
+
+        // $res->addListener($this);
+
+        foreach ($this->formatters as $formatter) {
+            $res->addListener($formatter);
+        }
+
+        /* Set PHPUnit error handler */
+        if ($this->useCustomErrorHandler) {
+            set_error_handler([$this, 'handleError'], E_ALL | E_STRICT);
+        }
+
+        $this->injectFilters($suite);
+        $suite->run($res);
+
+        foreach ($this->formatters as $formatter) {
+            $formatter->processResult($res);
+        }
+
+        /* Restore Phing error handler */
+        if ($this->useCustomErrorHandler) {
+            restore_error_handler();
+        }
+
+        $this->checkResult($res);
+    }
+
+    /**
+     * @param PHPUnit\Framework\TestSuite $suite
+     * @throws ReflectionException
+     */
+    private function injectFilters(PHPUnit\Framework\TestSuite $suite): void
+    {
+        $filterFactory = new PHPUnit\Runner\Filter\Factory();
+
+        if (empty($this->excludeGroups) && empty($this->groups)) {
+            return;
+        }
+
+        if (!empty($this->excludeGroups)) {
+            $filterFactory->addFilter(
+                new ReflectionClass(\PHPUnit\Runner\Filter\ExcludeGroupFilterIterator::class),
+                $this->excludeGroups
+            );
+        }
+
+        if (!empty($this->groups)) {
+            $filterFactory->addFilter(
+                new ReflectionClass(\PHPUnit\Runner\Filter\IncludeGroupFilterIterator::class),
+                $this->groups
+            );
+        }
+
+        $suite->injectFilter($filterFactory);
+    }
+
+    /**
+     * @param \PHPUnit\Framework\TestResult $res
+     */
+    private function checkResult(\PHPUnit\Framework\TestResult $res): void
+    {
+        if ($res->skippedCount() > 0) {
+            $this->hasSkipped = true;
+        }
+
+        if ($res->notImplementedCount() > 0) {
+            $this->hasIncomplete = true;
+        }
+
+        if ($res->warningCount() > 0) {
+            $this->hasWarnings = true;
+        }
+
+        if ($res->failureCount() > 0) {
+            $this->hasFailures = true;
+        }
+
+        if ($res->errorCount() > 0) {
+            $this->hasErrors = true;
+        }
+
+        if ($res->riskyCount() > 0) {
+            $this->hasRisky = true;
+        }
```

Comment:
Please reduce complexity and avoid not needed conditions.

`$this->hasFailures = $res->failureCount() > 0;`

As seen in the Codecov Report
> - Complexity     9119     9187      +68

Reply:
Change the complexity suggestion. The raise in missing coverage is likely due to adding PHPUnitTestRunner9 which results in PHPUnitTestRunner8 not being used.